nav2_costmap_2d/plugins/denoise_layer.cpp

```diff
@@ -0,0 +1,244 @@
+// Copyright (c) 2021 Andrey Ryzhikov
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_costmap_2d/denoise_layer.hpp"
+
+#include <string>
+#include <vector>
+#include <algorithm>
+
+#include "rclcpp/rclcpp.hpp"
+
+namespace nav2_costmap_2d
+{
+bool isBackground(uint8_t pixel)
+{
+  return pixel != LETHAL_OBSTACLE && pixel != NO_INFORMATION;
+}
+
+void
+DenoiseLayer::onInitialize()
+{
+  // Enable/disable plugin
+  declareParameter("enabled", rclcpp::ParameterValue(true));
+  // Smaller groups should be filtered
+  declareParameter("minimal_group_size", rclcpp::ParameterValue(2));
+  // Pixels connectivity type
+  declareParameter("group_connectivity_type", rclcpp::ParameterValue(8));
+  // Max additional memory amount
+  declareParameter("memory_usage_mb", rclcpp::ParameterValue(0));
```

Comment:
Some context here would be great - why is this a parameter? Shouldn't this be sized with the costmap?

Reply:
There is a trade-off between the amount of memory consumed by the plugin and the number of obstacle groups that can be segmented. Selecting type `uint32_t` for type of label consumes more memory (and slightly slower for this reason), but 2^32 groups of obstacles are allowed, which seems to be sufficient for costmap. Choosing type `uint16_t` cuts memory consumption in half, but this configuration only allows 2^16 obstacles.
In this regard, the type of label [is selected](https://github.com/ros-planning/navigation2/pull/2567/commits/ae847a7e548480895666eb07788245495f67be98#diff-102f2d1fcbf93fc1b8c041a37d7dbd6f5359c6f9616d28ef716ccffc9e5d5474R199) at runtime, depending on the amount of available memory. This parameter sets the maximum allowed memory size for the plugin. If it is omitted the memory will not be limited.

Reply:
2^16 is 65536, trying to backout a max size assuming they were all linearly separate with a 5cm costmap, that would be a 12 x 12m update size. Because for them to be different clusters there would have to be space between them, that would be closer to 24x24m. I agree, that's maybe a little small for me to consider it "good enough for all possible situations".

Would it be possible to attempt the 16 bit version and go to the 32 bit version only if it became a problem? Then probably cache it with a bool to then use 32 bit from there on out. I would just like to remove potentially confusing parameters that may be hard to explain fully if this is logic we can handle internally

Reply:
[Now](https://github.com/ros-planning/navigation2/pull/2567/commits/f79d6f2ffdcbe1ea004010f646fa6766a3e1ab4d) the switch uint16_t -> uint32_t is done automatically on overflow. Incomplete results of the calculation with uint16_t labels are discarded: I decided that transferring them is unreasonably difficult. The calculation simply restarts from scratch. For all subsequent calls, calculations will be made in uint32_t.